common/fixation_entry_test.go

```diff
@@ -0,0 +1,252 @@
+	require.NotNil(t, err)
+
+	// get the latest version and make sure it's equal to dummyObj2
+	err = vs.GetEntry(ctx, dummyIndex, blockToAddEntryAfterStale, &dummyCoin, types.DO_NOTHING)
+	require.Nil(t, err)
+	require.True(t, dummyCoin.IsEqual(dummyObj2))
+
+	// make sure dummy index is still in the entry index list
+	indexList = vs.GetAllEntryIndices(ctx)
+	require.Equal(t, 1, len(indexList))
+}
+
+// Test that when adds two entries with the same block and index and makes sure that only the latest one is kept
+func TestAdditionOfTwoEntriesWithSameIndexInSameBlock(t *testing.T) {
+	// create dummy data for two dummy entries
+	dummyIndex := "index"
+	dummyObj := sdk.Coin{Denom: "utest", Amount: sdk.ZeroInt()}
+	dummyObj2 := sdk.Coin{Denom: "utest", Amount: sdk.OneInt()}
+
+	// init FixationStore + context
+	vs, ctx := initCtxAndFixationStore(t)
+
+	// add the first dummy entry
+	blockToAddEntry := uint64(0)
+	err := vs.AppendEntry(ctx, dummyIndex, blockToAddEntry, &dummyObj)
+	require.Nil(t, err)
+
+	// add the second dummy entry
+	err = vs.AppendEntry(ctx, dummyIndex, blockToAddEntry, &dummyObj2)
+	require.Nil(t, err)
+
+	// get all entry indices and make sure there is only one index
+	indexList := vs.GetAllEntryIndices(ctx)
+	require.Equal(t, 1, len(indexList))
+
+	// get the entry from the storage
+	var dummyCoin sdk.Coin
+	err = vs.GetEntry(ctx, dummyIndex, blockToAddEntry, &dummyCoin, types.DO_NOTHING)
+	require.Nil(t, err)
+
+	// make sure that one entry's data is the same data of the second dummy entry
+	require.True(t, dummyCoin.IsEqual(dummyObj2))
+
+	// make sure dummy index is still in the entry index list
+	indexList = vs.GetAllEntryIndices(ctx)
+	require.Equal(t, 1, len(indexList))
+}
+
+// Test adding entry versions and getting an older version
+func TestEntryVersions(t *testing.T) {
+	// create dummy data for two dummy entries
+	dummyIndex := "index"
+	dummyObj := sdk.Coin{Denom: "utest", Amount: sdk.ZeroInt()}
+	dummyObj2 := sdk.Coin{Denom: "utest", Amount: sdk.OneInt()}
+
+	// init FixationStore + context
+	vs, ctx := initCtxAndFixationStore(t)
+
+	// add the first dummy entry
+	blockToAddFirstEntry := uint64(10)
+	err := vs.AppendEntry(ctx, dummyIndex, blockToAddFirstEntry, &dummyObj)
+	require.Nil(t, err)
+
+	// add the second dummy entry
+	blockToAddSecondEntry := uint64(20)
+	err = vs.AppendEntry(ctx, dummyIndex, blockToAddSecondEntry, &dummyObj2)
+	require.Nil(t, err)
+
+	// get the older version from block blockToAddFirstEntry
+	var dummyCoin sdk.Coin
+	err = vs.GetEntry(ctx, dummyIndex, blockToAddFirstEntry+1, &dummyCoin, types.DO_NOTHING)
+	require.Nil(t, err)
+
+	// verify the data matches the old entry from storage
+	require.True(t, dummyCoin.IsEqual(dummyObj))
+
+	// make sure dummy index is still in the entry index list
+	indexList := vs.GetAllEntryIndices(ctx)
+	require.Equal(t, 1, len(indexList))
+}
+
+// Test adding entry versions with different fixation keys
+func TestDifferentFixationKeys(t *testing.T) {
```

Comment:
test two entries being set at different heights, and removal only removes one of them

Reply:
done